core/src/main/java/org/apache/iceberg/BaseTable.java

```diff
@@ -74,6 +75,57 @@ public Schema schema() {
     return ops.current().schemasById();
   }
 
+  /**
+   * Return the {@link Schema schema} for this table at the time of the snapshot
+   * specified by the snapshotId.
+   *
+   * @return the schema
+   */
+  public Schema schemaForSnapshot(long snapshotId) {
```

Comment:
Now that snapshots can report the schema that was current when the snapshot was created, I think that the first PR should focus on using that schema and not recovering schemas from old metadata files. We may want to add this later, but it over-complicates getting the initial feature done.

Reply:
I will remove this and the following method from `BaseTable`.
I will move the recovery of schema from previous metadata files (in the event that a snapshot does not have a schema id associated with it) to a separate PR.